core/unit_test/TestGraph.hpp

```diff
@@ -349,14 +340,179 @@ TEST_F(TEST_CATEGORY_FIXTURE(graph), zero_work_reduce) {
+  graph.instantiate();
+  graph.submit(ex);
+  ex.fence();
+}
+
+template <typename ViewType, size_t TargetIndex, size_t NumIndices>
+struct AddTo {
+  ViewType data;
+  typename ViewType::value_type value;
+  Kokkos::Array<size_t, NumIndices> indices;
+
+  AddTo(typename ViewType::non_const_type data_,
+        typename ViewType::non_const_value_type value_,
+        Kokkos::Array<size_t, NumIndices> indices_)
+      : data(std::move(data_)), value(value_), indices(std::move(indices_)) {}
+
+  template <typename T>
+  KOKKOS_FUNCTION void operator()(const T) const {
+    if constexpr (NumIndices > 0) {
+      for (size_t index = 0; index < indices.size(); ++index)
+        data(TargetIndex) += data(indices[index]);
+    }
+    data(TargetIndex) += value;
+  }
+};
+
+// Ensure that we can handle the simple diamond use case.
+//
+// topology     stream-based approach       graph-based
+//
+//   A          A(exec_0)                   Using the API to add nodes, no
+//  / \         fence(exec_0)               user-facing fence anymore because
+// B   C        B(exec_0)   C(exec_1)       we'd like to rely on the graph to
+//  \ /         fence(exec_1)               enforce dependencies.
+//   D          D(exec_0)
+TEST_F(TEST_CATEGORY_FIXTURE(graph), diamond) {
+  const auto execution_space_instances =
+      Kokkos::Experimental::partition_space(ex, 1, 1);
+
+  const auto exec_0 = execution_space_instances.at(0);
+  const auto exec_1 = execution_space_instances.at(1);
+
+  using policy_t = Kokkos::RangePolicy<TEST_EXECSPACE>;
+  using view_t   = Kokkos::View<int*, TEST_EXECSPACE>;
+  using view_h_t = Kokkos::View<int*, Kokkos::HostSpace>;
+
+  view_t data(Kokkos::view_alloc(ex, "diamond - data"), 4);
+
+  constexpr int value_A = 42, value_B = 27, value_C = 13, value_D = 147;
+  constexpr int index_A = 0, index_B = 1, index_C = 2, index_D = 3;
+
+  auto graph = Kokkos::Experimental::create_graph(ex, [&](auto root) {
+    auto node_A = root.then_parallel_for(
+        policy_t(exec_0, 0, 1), AddTo<view_t, index_A, 0>(data, value_A, {}));
+
+    auto node_B = node_A.then_parallel_for(
+        policy_t(exec_0, 0, 1),
+        AddTo<view_t, index_B, 1>(data, value_B, {index_A}));
+    auto node_C = node_A.then_parallel_for(
+        policy_t(exec_1, 0, 1),
+        AddTo<view_t, index_C, 1>(data, value_C, {index_A}));
+
+    auto node_D =
+        Kokkos::Experimental::when_all(node_B, node_C)
+            .then_parallel_for(
+                policy_t(exec_0, 0, 1),
+                AddTo<view_t, index_D, 2>(data, value_D, {index_B, index_C}));
+  });
+  graph.instantiate();
+
+  // TODO Check that kernels are running on the execution space instance of
+  //      their policy if the defaulted graph implementation is used.
+  graph.submit(ex);
+
+  view_h_t data_host(
+      Kokkos::view_alloc(Kokkos::WithoutInitializing, "diamond - data - host"),
+      4);
+  Kokkos::deep_copy(ex, data_host, data);
+
+  ex.fence();
```

Comment:
Why not just `auto data_host =  create_mirror_view_and_copy(data);`

Reply:
I guess I just copied what was done in other tests of this file.

But also thinking of the test itself, I think it's a good idea to avoid any global fencing (that are implied by `create_mirror_view_and_copy` right ?). Because we want to ensure everything was properly enqueued.

Reply:
Sure